### DISCUSSION TITLE
Replacing `scope` in v2

### DISCUSSION THREAD

**Discussion by HonzaKirchner** [pmndrs/jotai#1880] [General] [answered]
Hey!

I was using scope in order to nest multiple JotaiProviders. The main idea was that all useAtom hooks would be subscribed to the global provider UNLESS the scope parameter would be used (I only had like 2 very specific hooks defined like that).
I was wondering if something like that is possible at all in v2? It seems that it would require updating all the `useAtom`s that are rendered within nested JotaiProvider and make them use the `store` parameter which is something I'd like to avoid.
 
<img width="686" alt="CleanShot 2023-04-12 at 21 08 59@2x" src="https://user-images.githubusercontent.com/17545432/231561113-1be687b0-14f3-4630-a019-3a1c514b0d77.png">

**Comment by dai-shi** [comment]
Check out the migration document if you haven't. https://jotai.org/docs/guides/migrating-to-v2-api

v1 scope was handy way to create React context dynamically. It was needed for library usage, and because we didn't expose store API.
v2 exposes store API, so scope isn't necessary. We can do the same thing with v2 API. (`initialValues` isn't 100% compatible, but it's not related only with `scope`.)


I think creating a compat layer is possible. Here's an idea:
```js
// jotai-compat.js for v1 compatible API
import { createContext, useContext, useRef } from 'react';
import * as JotaiV2 from 'jotai'; // jotai v2

const contexts = new Map();
const getContext = (scope) => {
  if (!contexts.has(scope)) {
    contexts.set(scope, createContext);
  }
  return contexts.get(scope);
};

export const Provider = ({ children, scope, initialValues = [] }) => {
  const Context = getContext(scope);
  const storeRef = useRef();
  if (!storeRef.current) {
    const store = JotaiV2.createStore();
    for (const [a, v] of initialValues) {
      store.set(a, v);
    }
    storeRef.current = store;
  }
  return (
    <Context.Provider value={storeRef.current}>{children}</Context.Provider>
  );
};

export const useAtom = (anAtom, scope) => {
  const Context = getContext(scope);
  const store = useContext(Context);
  return JotaiV2.useAtom(anAtom, { store });
};
```

However, in most cases, you don't need it. If you use infinite number of `scope`, it might be helpful.

**Comment by HonzaKirchner** [comment]
@dai-shi cool idea, thank you.

But how does the new API work for library authors? Let's say I have a library that leverages unmounting of provider to clean some atom data. 

<img width="330" alt="CleanShot 2023-04-16 at 20 48 27@2x" src="https://user-images.githubusercontent.com/17545432/232335051-c4e70ce9-e6b6-4cc1-9c4c-283d06fab3d4.png">

How can I do that without asking users of the library to create custom `useAtom` hook that would subscribe to theirs global store ( ignoring Providers that my library rendered).

**Reply by dai-shi** [reply]
> But how does the new API work for library authors?

I'm not sure if I get your concern. From the library perspective, v1 scope is 100% migrate-able to v2 createStore+createContext.

**Reply by HonzaKirchner** [reply]
I agree it's 100% migrate-able, but I don't see how it's done for library authors.
Let's say that as a library author, I want to offer a component that renders `Provider` without breaking users' app. If users of my library  also use Jotai on their own,  calling `useAtom` inside of my `Provider` won't get data from the global store but from the Provider that my library rendered (unless they deliberately use store parameter).
The thing is that I don't want to ask the users to start using store parameter. I would like to make their `useAtom` completely ignore my provider. That was possible in V1 by giving the provider some scope and I wonder if something like that is achievable in V2.

**Reply by dai-shi** [reply]
Ah, well, you as a library author should provide custom `Provider` and custom `useAtom`. Your library should hide Jotai completely. Users don't care if Jotai is used in it or not.

**Reply by HonzaKirchner** [reply]
And how do I actually hide it? If UI component from my library renders Provider internally, it's just a provider like any other isn't it? That's what my question is all about. Hiding it was easily achievable using `scope` param but I don't see an easy way to do it using `store`

**Reply by dai-shi** [reply]
> Hiding it was easily achievable using scope param but I don't see an easy way to do it using store

Hmm, I'm not sure if I understand how you used `scope`, but here's my perspective.

```js
// v1
const myAtom = atom(1);
const myScope = Symbol();
export function useMyState() {
  return useAtom(myAtom, myScope);
}

// v2
const myAtom = atom(1);
const MyContext = createContext();
export function useMyState() {
  const store = useContext(MyContext);
  return useAtom(myAtom, { store });
}
```

It's just a little lower level API. See closely [jotai-compat.js](https://github.com/pmndrs/jotai/discussions/1880#discussioncomment-5602530) if you haven't.

**Reply by HonzaKirchner** [reply]
Here is an [example](https://codesandbox.io/s/jotai-playground-forked-viye2o?file=/src/App.jsx) of how I used scope in V1.
Basically, the library users can just use Jotai and ignore that part of the tree is wrapped in another Provider. It's invisible to them because they don't know the scope name.

I checked [jotai-compat.js](https://github.com/pmndrs/jotai/discussions/1880#discussioncomment-5602530), but there, users would have to change their codebase and pass their global store to the `useAtom` call.

I know there is a way how to get data from any provider up the tree, but I don't see a way how to 'hide' the closest provider from `useAtom` calls that don't provide `store` parameter

**Reply by dai-shi** [reply, marked as answer]
Here's the migrated version: https://codesandbox.io/s/jotai-playground-forked-7kvov7?file=/src/App.jsx

**Comment by dai-shi** [comment]
FYI, we released a new package to mitigate the migration. https://github.com/jotaijs/jotai-scope Docs are not ready.
